<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>all [filter] 	Retrieves everthing
Specified [property] 	Formal binomial and trinomial
at or below species level [property]
family [rank] 	Rank-based query
taxonomy genome [filter] 	Taxa with a direct link to a genome sequence
2009/10/21:2020 [date] 	Date-bounded query
mammalia [subtree] 	All taxa within the Mammalia
extinct [property] 	Extinct organisms
Terminal [property] 	Terminal nodes in the tree
loprovencyclife [filter] 	Entries with LinkOut links to the Encyclopedia of Life — fetch_genbank • taxreturn</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="all [filter] 	Retrieves everthing
Specified [property] 	Formal binomial and trinomial
at or below species level [property]
family [rank] 	Rank-based query
taxonomy genome [filter] 	Taxa with a direct link to a genome sequence
2009/10/21:2020 [date] 	Date-bounded query
mammalia [subtree] 	All taxa within the Mammalia
extinct [property] 	Extinct organisms
Terminal [property] 	Terminal nodes in the tree
loprovencyclife [filter] 	Entries with LinkOut links to the Encyclopedia of Life — fetch_genbank" />
<meta property="og:description" content="Fetch sequences from genbank" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">taxreturn</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/taxreturn.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alexpiper/taxreturn/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>all <code>[filter]</code> 	Retrieves everthing
Specified <code>[property]</code> 	Formal binomial and trinomial
at or below species level <code>[property]</code>
family <code>[rank]</code> 	Rank-based query
taxonomy genome <code>[filter]</code> 	Taxa with a direct link to a genome sequence
2009/10/21:2020 <code>[date]</code> 	Date-bounded query
mammalia <code>[subtree]</code> 	All taxa within the Mammalia
extinct <code>[property]</code> 	Extinct organisms
Terminal <code>[property]</code> 	Terminal nodes in the tree
loprovencyclife <code>[filter]</code> 	Entries with LinkOut links to the Encyclopedia of Life</h1>
    <small class="dont-index">Source: <a href='https://github.com/alexpiper/taxreturn/blob/master/R/fetch_sequences.R'><code>R/fetch_sequences.R</code></a></small>
    <div class="hidden name"><code>fetch_genbank.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fetch sequences from genbank</p>
    </div>

    <pre class="usage"><span class='fu'>fetch_genbank</span><span class='op'>(</span>
  <span class='va'>x</span>,
  database <span class='op'>=</span> <span class='st'>"nuccore"</span>,
  marker <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"COI[GENE]"</span>, <span class='st'>"CO1[GENE]"</span>, <span class='st'>"COX1[GENE]"</span><span class='op'>)</span>,
  output <span class='op'>=</span> <span class='st'>"h"</span>,
  min_length <span class='op'>=</span> <span class='fl'>1</span>,
  max_length <span class='op'>=</span> <span class='fl'>2000</span>,
  subsample <span class='op'>=</span> <span class='cn'>FALSE</span>,
  chunk_size <span class='op'>=</span> <span class='fl'>100</span>,
  db <span class='op'>=</span> <span class='cn'>NULL</span>,
  multithread <span class='op'>=</span> <span class='cn'>FALSE</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  progress <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>A taxon name or vector of taxa to download sequences for</p></td>
    </tr>
    <tr>
      <th>database</th>
      <td><p>The database to download from. For NCBI GenBank this currently onlt accepts the arguments 'nuccore' or 'genbank' which is an alias for nuccore.</p></td>
    </tr>
    <tr>
      <th>marker</th>
      <td><p>The barcode marker used as a search term for the database.
If this is set to "mitochondria" or "mitochondrion" it will download full mitochondrial genomes. If set to "genome" it will download entire genomes only.</p></td>
    </tr>
    <tr>
      <th>output</th>
      <td><p>The output format for the taxonomy in fasta headers.
Options include "h" for full heirarchial taxonomy (Accession;Domain;Phylum;Class;Order;Family;Genus;Species),
"binom" for just genus species binomials (Accession;Genus_species),
"bold" for BOLD taxonomic ID only (Accession;BoldTaxID),
"gb" for genbank taxonomic ID (Accession|GBTaxID),
"gb-binom" which outputs genbank taxonomic ID's and Genus species binomials, translating BOLD taxonomic ID's to genbank in the process (Accession|GBTaxID;Genus_species)
or "standard" which outputs the default format for each database. For bold this is <code>sampleid|species name|markercode|genbankid</code></p></td>
    </tr>
    <tr>
      <th>min_length</th>
      <td><p>The minimum length of sequences to download</p></td>
    </tr>
    <tr>
      <th>max_length</th>
      <td><p>The maximum length of sequences to download</p></td>
    </tr>
    <tr>
      <th>subsample</th>
      <td><p>(Numeric) return a random subsample of sequences from the search.</p></td>
    </tr>
    <tr>
      <th>chunk_size</th>
      <td><p>Split up the query into chunks of this size to avoid overloading API servers. if left NULL, the default will be 300</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p>a database file generated using <code><a href='get_ncbi_taxonomy.html'>taxreturn::get_ncbi_taxonomy()</a></code>. Generated automatically if NULL.</p></td>
    </tr>
    <tr>
      <th>multithread</th>
      <td><p>Whether multithreading should be used, if TRUE the number of cores will be automatically detected, or provided a numeric vector to manually set the number of cores to use</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>Whether progress should be printed to the console.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p>A logical, for whether or not to print a progress bar when multithread is true. Note, this will slow down processing.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    



  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Alexander Piper.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


