<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ taxreturn</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">taxreturn</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/taxreturn.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alexpiper/taxreturn/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="DNAbin2DNAstringset.html">DNAbin2DNAstringset()</a></code> </p>
        </td>
        <td><p>Convert DNABin to DNAStringSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DNAbin2char.html">DNAbin2char()</a></code> </p>
        </td>
        <td><p>DNAbin to char</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="acc2hex.html">acc2hex()</a></code> </p>
        </td>
        <td><p>Convert accession number to hexadecimal coding</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="acc_from_fasta.html">acc_from_fasta()</a></code> </p>
        </td>
        <td><p>Genbank accessions from fasta</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="alignment_entropy.html">alignment_entropy()</a></code> </p>
        </td>
        <td><p>Alignment entropy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast.html">blast()</a></code> </p>
        </td>
        <td><p>Run BLAST search</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_assign_species.html">blast_assign_species()</a></code> </p>
        </td>
        <td><p>Assign species using BLAST</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_install.html">blast_install()</a></code> </p>
        </td>
        <td><p>Install BLAST</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_params.html">blast_params()</a></code> </p>
        </td>
        <td><p>Show BLAST parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_top_hit.html">blast_top_hit()</a></code> </p>
        </td>
        <td><p>BLAST Top Hit</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="char2DNAbin.html">char2DNAbin()</a></code> </p>
        </td>
        <td><p>char2DNAbin</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clean_seqs.html">clean_seqs()</a></code> </p>
        </td>
        <td><p>Clean Sequences with PHMM (DEPRECATED)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="closest_seq.html">closest_seq()</a></code> </p>
        </td>
        <td><p>Closest kmer distance
This function checks the kmer distance between all the sequences in a DNAbin and a reference PHMM model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="codon_entropy.html">codon_entropy()</a></code> </p>
        </td>
        <td><p>Codon entropy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="codon_filter.html">codon_filter()</a></code> </p>
        </td>
        <td><p>Filter sequences containing stop codons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="concat_DNAbin.html">concat_DNAbin()</a></code> </p>
        </td>
        <td><p>Concatenate DNAbin objects while preserving attributes.
This function joins two or more <code>DNAbin</code> objects, retaining any
attributes whose lengths match those of the input objects (e.g. "species",
"lineage" and/or "taxID" attributes).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-findExecutable.html">.findExecutable()</a></code> </p>
        </td>
        <td><p>Find executable</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-hex.html">.hex()</a></code> </p>
        </td>
        <td><p>convert alphanumerics to hexadecimal
internal taxreturn function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-ishex.html">.ishex()</a></code> </p>
        </td>
        <td><p>Check whether a string is hexadecimal
internal taxreturn function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-unhex.html">.unhex()</a></code> </p>
        </td>
        <td><p>convert hexadecimal strings to alphanumeric
internal taxreturn function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="download_ott_taxonomy.html">download_ott_taxonomy()</a></code> </p>
        </td>
        <td><p>Download open tree of life taxonomy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetchSeqs.html">fetchSeqs()</a></code> </p>
        </td>
        <td><p>Fetchseqs function (DEPRECATED)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_bold.html">fetch_bold()</a></code> </p>
        </td>
        <td><p>Fetch sequences from BOLD</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_genbank.html">fetch_genbank()</a></code> </p>
        </td>
        <td><p>all <code>[filter]</code> 	Retrieves everthing
Specified <code>[property]</code> 	Formal binomial and trinomial
at or below species level <code>[property]</code>
family <code>[rank]</code> 	Rank-based query
taxonomy genome <code>[filter]</code> 	Taxa with a direct link to a genome sequence
2009/10/21:2020 <code>[date]</code> 	Date-bounded query
mammalia <code>[subtree]</code> 	All taxa within the Mammalia
extinct <code>[property]</code> 	Extinct organisms
Terminal <code>[property]</code> 	Terminal nodes in the tree
loprovencyclife <code>[filter]</code> 	Entries with LinkOut links to the Encyclopedia of Life</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_seqs.html">fetch_seqs()</a></code> </p>
        </td>
        <td><p>fetch_seqs function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_by_tax.html">filter_by_tax()</a></code> </p>
        </td>
        <td><p>Filter sequences by taxonomy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_infraspecifc.html">filter_infraspecifc()</a></code> </p>
        </td>
        <td><p>Filter infraspecific taxonomic labels</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_unplaced.html">filter_unplaced()</a></code> </p>
        </td>
        <td><p>Filter unplaced taxonomic labels</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_binding_position.html">get_binding_position()</a></code> </p>
        </td>
        <td><p>Get primer binding position</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_mixed_clusters.html">get_mixed_clusters()</a></code> </p>
        </td>
        <td><p>Get mixed clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ncbi_lineage.html">get_ncbi_lineage()</a></code> </p>
        </td>
        <td><p>Get lineage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ncbi_synonyms.html">get_ncbi_synonyms()</a></code> </p>
        </td>
        <td><p>Get NCBI synonyms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ncbi_taxonomy.html">get_ncbi_taxonomy()</a></code> </p>
        </td>
        <td><p>Download NCBI taxdump</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ott_lineage.html">get_ott_lineage()</a></code> </p>
        </td>
        <td><p>Recursively get lineage from OTT taxid
This function derives the full lineage of a taxon ID number from a given taxonomy database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ott_taxonomy.html">get_ott_taxonomy()</a></code> </p>
        </td>
        <td><p>get_ott_taxonomy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_reading_frame.html">get_reading_frame()</a></code> </p>
        </td>
        <td><p>Get Reading frame of sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_subalignment.html">get_subalignment()</a></code> </p>
        </td>
        <td><p>Get subalignment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gid_to_acc.html">gid_to_acc()</a></code> </p>
        </td>
        <td><p>Convert NCBI gene ids to accession numbers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graft_tips.html">graft_tips()</a></code> </p>
        </td>
        <td><p>Graft new tips onto an existing tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hex2acc.html">hex2acc()</a></code> </p>
        </td>
        <td><p>Hexadecimal coding to accession numberD</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kmer_screen.html">kmer_screen()</a></code> </p>
        </td>
        <td><p>Kmer screening function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lca_probs.html">lca_probs()</a></code> </p>
        </td>
        <td><p>Probabilitys of sharing a rank as a function of sequence identity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_blast_db.html">make_blast_db()</a></code> </p>
        </td>
        <td><p>Make blast Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeblastdb.html">makeblastdb()</a></code> </p>
        </td>
        <td><p>Make blast Database (DEPRECATED)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_to_model.html">map_to_model()</a></code> </p>
        </td>
        <td><p>Map to model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_to_ott.html">map_to_ott()</a></code> </p>
        </td>
        <td><p>Map taxa to open tree of life</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="model.html">model</a></code> </p>
        </td>
        <td><p>COI phmm model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ncbi_taxid.html">ncbi_taxid()</a></code> </p>
        </td>
        <td><p>Get ncbi taxid's for a taxon name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pad_alignment.html">pad_alignment()</a></code> </p>
        </td>
        <td><p>Pad alignment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_gb.html">parse_gb()</a></code> </p>
        </td>
        <td><p>Parse genbank flat files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_ott_synonyms.html">parse_ott_synonyms()</a></code> </p>
        </td>
        <td><p>parse the open tree of life synonyms file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="propagate_tax.html">propagate_tax()</a></code> </p>
        </td>
        <td><p>Propagate taxonomy (DEPRECATED)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prune_groups.html">prune_groups()</a></code> </p>
        </td>
        <td><p>Prune group sizes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_seq.html">random_seq()</a></code> </p>
        </td>
        <td><p>Generate random sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_bold_chunk.html">read_bold_chunk()</a></code> </p>
        </td>
        <td><p>read bold chunk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_genbank_chunk.html">read_genbank_chunk()</a></code> </p>
        </td>
        <td><p>Read genbank chunk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reformat_dada2_gen.html">reformat_dada2_gen()</a></code> </p>
        </td>
        <td><p>Reformat DADA2 Genus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reformat_dada2_spp.html">reformat_dada2_spp()</a></code> </p>
        </td>
        <td><p>Reformat DADA2 Species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reformat_hierarchy.html">reformat_hierarchy()</a></code> </p>
        </td>
        <td><p>Reformat hierarchy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resolve_synonyms_ncbi.html">resolve_synonyms_ncbi()</a></code> </p>
        </td>
        <td><p>resolve_synonyms_ncbi</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setup_multithread.html">setup_multithread()</a></code> </p>
        </td>
        <td><p>Setup multithreading</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setup_para.html">setup_para()</a></code> </p>
        </td>
        <td><p>Setup parallel</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_bold_query.html">split_bold_query()</a></code> </p>
        </td>
        <td><p>Split bold query
This function recursively splits a bold taxonomic query until the amount of records returned is under chunk_size</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_model.html">subset_model()</a></code> </p>
        </td>
        <td><p>Subset PHMM
Modified from Cameron Nugent's COIL package https://github.com/CNuge/coil/blob/master/R/subset_PHMM.r</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarise_fasta.html">summarise_fasta()</a></code> </p>
        </td>
        <td><p>summarise_fasta</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tax2phylo.html">tax2phylo()</a></code> </p>
        </td>
        <td><p>tax2phylo</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tax2tree.html">tax2tree()</a></code> </p>
        </td>
        <td><p>tax2tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="train_idtaxa.html">train_idtaxa()</a></code> </p>
        </td>
        <td><p>Train IDTAXA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_fasta.html">write_fasta()</a></code> </p>
        </td>
        <td><p>write fasta</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Alexander Piper.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


